pkg/controller/deployment/deployment_controller.go

```diff
@@ -564,6 +567,58 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// proportionalScaling should run only on scaling events and not during the normal rollout process.
+func (dc *DeploymentController) proportionalScaling(deployment extensions.Deployment) error {
+	newRS, oldRSs, err := dc.getAllReplicaSets(deployment)
+	if err != nil {
+		return err
+	}
+	if len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+		// No old replica sets with pods; scale new replica set to the full count.
+		_, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment)
+		return err
+	}
+	allRSs := append(controller.FilterActiveReplicaSets(oldRSs), newRS)
+
+	sort.Sort(controller.ByAvailableReplicas(allRSs))
+
+	maybeMaxSurge := 0
+	if deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType {
+		maybeMaxSurge, err = intstrutil.GetValueFromIntOrPercent(&deployment.Spec.Strategy.RollingUpdate.MaxSurge, deployment.Spec.Replicas)
+		if err != nil {
+			return err
+		}
+	}
+
+	totalReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+	// Number of additional pods that can be either added or removed from the total pod count.
+	additional := deployment.Spec.Replicas + maybeMaxSurge - totalReplicas
+	if additional == 0 {
+		return nil
+	}
+	// Scale proportionally
+	for i := range allRSs {
+		rs := allRSs[i]
+		// This can be either positive (scale up) or negative (scale down).
+		// TODO: Rounding and distributing of pods need more care
+		proportion := int((rs.Status.Replicas / totalReplicas) * additional)
```

Comment:
@bgrant0607 note that rs.status.replicas doesn't mean available pods but [running](https://github.com/kubernetes/kubernetes/blob/524aaa2f7f639e534d6741853f5700b12fb929f5/pkg/controller/controller_utils.go#L419) pods. Those pods may be unready eg. due to failed probes.